File: ocaml/robot-name/robot_name.ml
type robot = { mutable name : string }

let int_in_range min max = min + Random.int (max - min + 1)
let char_in_range min max = Char.(int_in_range (code min) (code max) |> chr)

let generate_random_name () =
  Random.self_init ();
  let c () = char_in_range 'A' 'Z' and d () = char_in_range '0' '9' in
  [ c; c; d; d; d ] |> List.map (fun f -> f ()) |> List.to_seq |> String.of_seq

module StringSet = Set.Make (String)

let used_names = ref StringSet.empty

let rec generate_unique_name () =
  let new_name = generate_random_name () in
  if StringSet.mem new_name !used_names then
    (generate_unique_name [@tailcall]) ()
  else (
    used_names := StringSet.add new_name !used_names;
    new_name)

let new_robot () = { name = generate_unique_name () }
let name x = x.name
let reset x = x.name <- generate_unique_name ()

File: cpp/robot-name/robot_name.cpp
#include "robot_name.h"

namespace robot_name {

std::unordered_set<std::string> robot::used_names;

std::string robot::generate_unique_name() {
  std::string result;

  do {
    result = "";
    result.push_back('A' + std::rand() % 26);
    result.push_back('A' + std::rand() % 26);
    result.push_back('0' + std::rand() % 10);
    result.push_back('0' + std::rand() % 10);
    result.push_back('0' + std::rand() % 10);
  } while (robot::used_names.find(result) != robot::used_names.end());

  robot::used_names.insert(result);

  return result;
}

robot::robot() { m_name = generate_unique_name(); }

std::string robot::name() const { return m_name; }

void robot::reset() { m_name = generate_unique_name(); }

} // namespace robot_name
